config_v2.md

```diff
@@ -0,0 +1,370 @@
+    nb_epochs=ConfigValue(50, "The number of epochs")
+))
+print(configuration['batch_size'])  # will print 32, not a ConfigValue.
+
+ex = sacred.Experiment('my_pretty_experiment', config=configuration)
+                       
+def my_main_function(batch_size, dataset_size, nb_epochs):
+    # main experiment here
+    ...
+
+with ex.start():
+    my_main_function(**configuration)
+```
+
+
+### Example with dynamic population of the potential modifications.
+
+
+There is a hierarchy now. Similar to ingredients.
+
+```python
+import sacred
+from sacred import Config
+
+from somewhere_else import get_mnist, get_cifar
+
+
+def config_mnist1(config):
+    config['dataset_args'] = dict(crop_size=(30, 30), random_flip=True)
+
+
+def config_mnist2(config):
+    config['dataset_args'] = dict(crop_size=(40, 60), random_flip=False)
+
+
+def config_dataset_mnist(config):
+    config['function_get_dataset'] = get_mnist
+    config.add_potential_modification(config_mnist1)
+    config.add_potential_modification(config_mnist2)
+    
+
+
+def config_cifar1(config):
+    config['dataset_args'] = dict(color=True, random_flip=True)
+
+
+def config_cifar2(config):
+    config['dataset_args'] = dict(color=False, random_flip=False)
+
+
+def config_dataset_cifar(config):
+    config['function_get_dataset'] = get_cifar
+    config.add_potential_modification(config_cifar1)
+    config.add_potential_modification(config_cifar2)
+
+
+potential_modifs = [config_dataset_mnist, config_dataset_cifar]
+configuration = Config(dict(dataset_size=10_000, nb_epochs=50),
+                       potential_modifications=potential_modifs)
+
+
+ex = sacred.Experiment('my_pretty_experiment', config=configuration)
+                       
+def my_main_function(dataset_size, nb_epochs, function_get_dataset, dataset_args):
+    # main experiment here
+    
+    my_dataset = function_get_dataset(**dataset_args)
+    my_dataset = my_dataset[:dataset_size]
+    ...
+
+with ex.start():
+    my_main_function(**configuration)
+```
+
+From the command line:
+
+```bash
+python my_main.py with config_dataset_cifar config_cifar2
+```
+
+
+### Example with default parameters.
+
+We'll provide a function `sacred.get_default_args(some_function)` which will extract the default arguments and their values to make a dictionary. The you should be able to use that in the configs. 
+
+
+
+```python
+import sacred
+from sacred import Config, get_default_args
+
+
+def get_mnist(crop_size, random_flip=True, brightness=0.5):
+    ...
+    
+print(get_default_args(get_mnist))
+# {'random_flip': True, 'brightness': 0.5}
+
+def config_mnist1(config):
+    config['dataset_args'].update(dict(crop_size=(30, 30), random_flip=True))
+
+
+def config_mnist2(config):
+    config['dataset_args'].update(dict(crop_size=(40, 60), random_flip=False))
+
+
+def config_dataset_mnist(config):
+    config['function_get_dataset'] = get_mnist
+    config['dataset_args'] = get_default_args(get_mnist)
```

Comment:
This is nice -- I quite often find myself with duplicate default parameter values in Sacred config and a function.

Reply:
Thank you! It makes sense for the default values in the config to be the default arguments of the function.